swarm/shed/index.go

```diff
@@ -0,0 +1,264 @@
+		i.AccessTimestamp = i2.AccessTimestamp
+	}
+	if i.StoreTimestamp == 0 {
+		i.StoreTimestamp = i2.StoreTimestamp
+	}
+	if i.UseMockStore == nil {
+		i.UseMockStore = i2.UseMockStore
+	}
+	return i
+}
+
+// Index represents a set of LevelDB key value pairs that have common
+// prefix. It holds functions for encoding and decoding keys and values
+// to provide transparent actions on saved data which inclide:
+// - getting a particular IndexItem
+// - saving a particular IndexItem
+// - iterating over a sorted LevelDB keys
+// It implements IndexIteratorInterface interface.
+type Index struct {
+	db              *DB
+	prefix          []byte
+	encodeKeyFunc   func(fields IndexItem) (key []byte, err error)
+	decodeKeyFunc   func(key []byte) (e IndexItem, err error)
+	encodeValueFunc func(fields IndexItem) (value []byte, err error)
+	decodeValueFunc func(value []byte) (e IndexItem, err error)
+}
+
+// IndexFuncs structure defines functions for encoding and decoding
+// LevelDB keys and values for a specific index.
+type IndexFuncs struct {
+	EncodeKey   func(fields IndexItem) (key []byte, err error)
+	DecodeKey   func(key []byte) (e IndexItem, err error)
+	EncodeValue func(fields IndexItem) (value []byte, err error)
+	DecodeValue func(value []byte) (e IndexItem, err error)
+}
+
+// NewIndex returns a new Index instance with defined name and
+// encoding functions. The name must be unique and will be validated
+// on database schema for a key prefix byte.
+func (db *DB) NewIndex(name string, funcs IndexFuncs) (f Index, err error) {
+	id, err := db.schemaIndexPrefix(name)
+	if err != nil {
+		return f, err
+	}
+	prefix := []byte{id}
+	return Index{
+		db:     db,
+		prefix: prefix,
+		// This function adjusts Index LevelDB key
+		// by appending the provided index id byte.
+		// This is needed to avoid collisions between keys of different
+		// indexes as all index ids are unique.
+		encodeKeyFunc: func(e IndexItem) (key []byte, err error) {
+			key, err = funcs.EncodeKey(e)
+			if err != nil {
+				return nil, err
+			}
+			return append(append(make([]byte, 0, len(key)+1), prefix...), key...), nil
+		},
+		// This function reverses the encodeKeyFunc constructed key
+		// to transparently work with index keys without their index ids.
+		// It assumes that index keys are prefixed with only one byte.
+		decodeKeyFunc: func(key []byte) (e IndexItem, err error) {
+			return funcs.DecodeKey(key[1:])
+		},
+		encodeValueFunc: funcs.EncodeValue,
+		decodeValueFunc: funcs.DecodeValue,
+	}, nil
+}
+
+// Get accepts key fields represented as IndexItem to retrieve a
+// value from the index and return maximum available information
+// from the index represented as another IndexItem.
+func (f Index) Get(keyFields IndexItem) (out IndexItem, err error) {
+	key, err := f.encodeKeyFunc(keyFields)
+	if err != nil {
+		return out, err
+	}
+	value, err := f.db.Get(key)
+	if err != nil {
+		return out, err
+	}
+	out, err = f.decodeValueFunc(value)
```

Comment:
shouldn't the filling of the values in `Join` be handled in the `decodeValueFunc`? (it would require sending the key too, obviously

Reply:
Actually not, as for the exact reason you mentioned. We would require that decodeValueFunc and decodeKeyFunc are correlated.